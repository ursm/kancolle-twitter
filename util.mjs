import base64url from 'base64url'
import crypto from 'crypto'
import querystring from 'querystring'

export function createFlecktarnUrl(url, {url: rootURL, hmacSecret}) {
  const signature = base64url.encode(crypto.createHmac('sha224', hmacSecret).update(url).digest())

  return `${rootURL}/images/${signature}/${querystring.escape(url)}`
}
