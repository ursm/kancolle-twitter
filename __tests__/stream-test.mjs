import EventEmitter from 'events'
import Twitter from 'twitter-lite'
import fetch from 'node-fetch'

import createStream from '../stream.mjs'
import { createFlecktarnUrl } from '../util.mjs'

const mockStream = jest.fn().mockImplementation(() => new EventEmitter())

jest.mock('twitter-lite', () => {
  return jest.fn().mockImplementation(() => {
    return {
      stream: mockStream
    }
  })
})

jest.mock('node-fetch')

const config = {
  idobata: {
    hookUrl: 'http://example.com/idobata-hook'
  },
  twitter: {
    followIds: ['FOLLOW_1', 'FOLLOW_2']
  },
  flecktarn: {
    rootUrl:    'http://example.com/flecktarn',
    hmacSecret: 'SECRET'
  }
}

let stream

beforeEach(() => {
  Twitter.mockClear()
  fetch.mockClear()
  mockStream.mockClear()

  stream = createStream(config)
})

test('connect', () => {
  expect(mockStream).toHaveBeenCalledWith('statuses/filter', {follow: 'FOLLOW_1,FOLLOW_2'})
})

describe('on tweet', () => {
  test('anyone else', () => {
    stream.emit('data', {
      user: {
        id_str: 'SOMEBODY_ID'
      }
    })

    expect(fetch).not.toHaveBeenCalled()
  })

  test('simple', () => {
    stream.emit('data', {
      user: {
        id_str: 'FOLLOW_1',
        profile_image_url_https: 'http://example.com/profile.png',
        name: 'Alice Liddell',
        screen_name: 'alice',
      },
      id_str: 'TWEET_ID',
      text: "hello from\nthe wonderland",
      extended_entities: {
        media: [
          {
            type: 'photo',
            media_url_https: 'http://example.com/photo/1.jpg',
            expanded_url: 'http://example.com/photo/1',
            sizes: {
              medium: {
                w: 1,
                h: 2
              }
            }
          },
          {
            type: 'video',
            media_url_https: 'http://example.com/video-thumbnail/1.jpg',
            expanded_url: 'http://example.com/video/1',
            sizes: {
              medium: {
                w: 3,
                h: 4
              }
            }
          }
        ]
      }
    })

    expect(fetch).toHaveBeenCalledTimes(1)

    const [url, {method, headers, body}] = fetch.mock.calls[0]
    const {source, format} = JSON.parse(body)

    expect(url).toBe('http://example.com/idobata-hook')
    expect(method).toBe('post')
    expect(headers).toEqual({'Content-Type': 'application/json'})
    expect(format).toBe('html')

    expect(source).toMatchSnapshot()
  })

  test('extended', () => {
    stream.emit('data', {
      user: {
        id_str: 'FOLLOW_2',
        profile_image_url_https: 'http://example.com/profile.png',
        name: 'Alice Liddell',
        screen_name: 'alice',
      },
      id_str: 'TWEET_ID',
      text: 'hello from the wonderland',
      extended_tweet: {
        full_text: 'Alice taking "Drink Me" bottle',
        display_text_range: [13, 23],
        extended_entities: {
          media: [
            {
              type: 'photo',
              media_url_https: 'http://example.com/photo/1.jpg',
              expanded_url: 'http://example.com/photo/1',
              sizes: {
                medium: {
                  w: 1,
                  h: 2
                }
              }
            },
            {
              type: 'video',
              media_url_https: 'http://example.com/video-thumbnail/1.jpg',
              expanded_url: 'http://example.com/video/1',
              sizes: {
                medium: {
                  w: 3,
                  h: 4
                }
              }
            }
          ]
        }
      }
    })

    expect(fetch).toHaveBeenCalledTimes(1)

    const [, {body}] = fetch.mock.calls[0]
    const {source} = JSON.parse(body)

    expect(source).toMatchSnapshot()
  })
})

describe('on delete', () => {
  test('ignore', () => {
    stream.emit('data', {
      delete: {
      }
    })

    expect(fetch).not.toHaveBeenCalled()
  })
})
