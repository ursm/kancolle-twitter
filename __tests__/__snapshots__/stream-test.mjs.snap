// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on tweet extended 1`] = `
"
  <img height=\\"16\\" width=\\"16\\" src=\\"http://example.com/profile.png\\">
  <b>Alice Liddell</b> (<a href=\\"https://twitter.com/alice\\">@alice</a>)<br>

  <p>
      \\"Drink Me\\"

    (<a href=\\"https://twitter.com/alice/status/TWEET_ID\\">link</a>)
  </p>

    <ul class=\\"list-inline\\">
        <li>
          <a href=\\"http://example.com/photo/1\\">
            <img src=\\"http://example.com/flecktarn/sr9FHUzSnLCFqNKFQF0XxTnt945OZ7T8QThH8A/http%3A%2F%2Fexample.com%2Fphoto%2F1.jpg\\" alt=\\"\\">
          </a>
        </li>
        <li>
          <a href=\\"http://example.com/video/1\\">
            <img src=\\"http://example.com/flecktarn/OrZAmsZyZb_39gAn4-NSiY24MerfSvK1ymIxIQ/http%3A%2F%2Fexample.com%2Fvideo-thumbnail%2F1.jpg\\" alt=\\"\\">
          </a>
        </li>
    </ul>"
`;

exports[`on tweet only entities 1`] = `
"
  <img height=\\"16\\" width=\\"16\\" src=\\"http://example.com/profile.png\\">
  <b>Alice Liddell</b> (<a href=\\"https://twitter.com/alice\\">@alice</a>)<br>

  <p>
      hello from the wonderland

    (<a href=\\"https://twitter.com/alice/status/TWEET_ID\\">link</a>)
  </p>

    <ul class=\\"list-inline\\">
        <li>
          <a href=\\"http://example.com/photo/1\\">
            <img src=\\"http://example.com/flecktarn/sr9FHUzSnLCFqNKFQF0XxTnt945OZ7T8QThH8A/http%3A%2F%2Fexample.com%2Fphoto%2F1.jpg\\" alt=\\"\\">
          </a>
        </li>
    </ul>"
`;

exports[`on tweet simple 1`] = `
"
  <img height=\\"16\\" width=\\"16\\" src=\\"http://example.com/profile.png\\">
  <b>Alice Liddell</b> (<a href=\\"https://twitter.com/alice\\">@alice</a>)<br>

  <p>
      hello from the wonderland

    (<a href=\\"https://twitter.com/alice/status/TWEET_ID\\">link</a>)
  </p>"
`;
